### ISSUE TITLE
Bundler 2.2.3 introduced a significant performance regression for our bundle, not fixed in 2.2.5

### ISSUE THREAD

**Issue by haines** [rubygems/rubygems#4275]
### Describe the problem as clearly as you can

Our `bundle install` (113 Gemfile dependencies, 253 gems) is very slow on recent versions of Bundler.

| Bundler version | Install time |
|---|---|
| 2.1.4 | 0.69s |
| 2.2.0 | 1.08s |
| 2.2.1 | 1.15s |
| 2.2.2 | 1.05s |
| 2.2.3 | 33.61s |
| 2.2.4 | 32.88s |
| 2.2.5 | 32.07s |

### Post steps to reproduce the problem

I don't want to disclose our full dependencies too widely but can add maintainers to a private repo on an individual basis.

### Which command did you run?

`bundle install`

### What were you expecting to happen?

Gems to install in a reasonable time.

### What actually happened?

Gems took a very long time to install.

#### Environment

```
Bundler       2.2.5
  Platforms   ruby, x86_64-darwin-19
Ruby          2.7.2p137 (2020-10-01 revision 5445e0435260b449decf2ac16f9d09bae3cafe72) [x86_64-darwin19]
  Full Path   /Users/andrew/.rubies/ruby-2.7.2/bin/ruby
  Config Dir  /Users/andrew/.rubies/ruby-2.7.2/etc
RubyGems      3.2.5
  Gem Home    /Users/andrew/.gem/ruby/2.7.2
  Gem Path    /Users/andrew/.gem/ruby/2.7.2:/Users/andrew/.rubies/ruby-2.7.2/lib/ruby/gems/2.7.0
  User Home   /Users/andrew
  User Path   /Users/andrew/.gem/ruby/2.7.0
  Bin Dir     /Users/andrew/.gem/ruby/2.7.2/bin
Tools         
  Git         2.30.0
  RVM         not installed
  rbenv       not installed
  chruby      0.3.9
```

#### Bundler Build Metadata

```
Built At          2021-01-11
Git SHA           ba867aed7f
Released Version  true
```

**Comment by deivid-rodriguez** [comment]
Thanks for the report @haines, could you add me to your private repo I guess?

**Comment by haines** [comment]
Done! Thanks for looking into it @deivid-rodriguez ❤️

**Comment by deivid-rodriguez** [comment]
@haines I still haven't figured out the regression, but I discovered that the regression is not in the resolution phase but in the materialization phase, and I resurrected some improvements I had done some time ago in that area in #4277.

Since I'm getting quite different timings from the ones in your chart, would you be able to verify the actual impact of the changes in your case? I expect a 4x improvement, so it should go down to ~8 seconds.

**Comment by haines** [comment]
Yep, even better in fact - I tested with c858dfc7adc9b49299383e645620b5243f4e62d0 and the time dropped to ~2.5 seconds!

**Comment by deivid-rodriguez** [comment]
That's great! I'm currently working on another potential speedup on the same area that should cut another 50%, evaluating feasibility right now. Then I'll get to the actual regression :laughing:.

**Comment by deivid-rodriguez** [comment]
Ok @haines, I figured out the regression. Could you try the latest commit in #4277 now? I've still pending the extra 50% improvement that I was evaluating but it's tricky. Let's see how this works for you.

**Comment by haines** [comment]
Yup, with 9c12e5f8c4da1089ecdf4b9ea8bbce9385c053c9 we're back down to 0.68s, the same as 2.1.4 🎉

**Comment by deivid-rodriguez** [comment]
That's awesome @haines! I'll split the regression fix and the extra speed up into separate PR and try to add some specs :+1:.

**Comment by ayanko** [comment]
@deivid-rodriguez 

I also can confirm that since **2.2.3** simple `bundle install` (even with lock file) is VERY VERY slow (at least 5 times).
And **2.2.6** release does NOT fix this issue.

But it works again FAST when I apply this commit https://github.com/rubygems/rubygems/commit/9c12e5f8c4da1089ecdf4b9ea8bbce9385c053c9

**Comment by deivid-rodriguez** [comment]
@haines I finally managed to figure out the extra 50% speed up I was working on. Can you try #4289? In my case it reduces `bundle install` time from 28s to 1s.

**Comment by haines** [comment]
Yep, I get a similar reduction from 5c312ddc37128e2af96c5171b1a61b7e15db0646 👏

**Comment by deivid-rodriguez** [comment]
Grea, thanks @haines! I'm sorry this is taking way more than I was expecting. I just don't want to cause more regressions :sweat_smile:.

Hopefully next week.

**Comment by bquorning** [comment]
In my project, `bundle check` increased from 0.55 seconds in Bundler v2.2.2 to around 7.0 seconds using Bundler v2.2.3–v2.2.6.

Using master branch (b5a113c040a1820a9f0fdd9a11965e695939e2be, after #4289 was merged) `bundle check` still takes around 7.0 seconds.

Using 9c12e5f8c4da1089ecdf4b9ea8bbce9385c053c9 I can run `bundle check` in 0.44 seconds 🎉

**Comment by deivid-rodriguez** [comment]
Yes, in the end I split out the fix for this issue from #4289, that's why this issue remained opened. The reason is that it caused some performance decrease in other cases. I'm currently working on this and hope to get this fixed by the end of the week.
